### DISCUSSION TITLE
Feature discussion: lazy build (no build step)

### DISCUSSION THREAD

**Discussion by orenelbaum** [solidjs/solid-start#172] [Ideas]
Two frameworks I've seen that have this feature of having "no build step":
https://ultrajs.dev/
https://fresh.deno.dev/
I'd rather call this lazy or runtime build cause we are still building our code we just don't do it ahead of time.
I'm not really sure about the benefits/tradeoffs and whether it would make sense in Solid Start, which is why I'm starting this discussion.

**Comment by kinthaiofficial** [comment]
Lazy build (no build step) is a fascinating approach — it's essentially what Vite does in dev mode, and extending that to production is the cutting edge of the bundler space.

The tension:

**Why build steps exist**:
- Bundle splitting for optimal load performance
- Tree shaking (eliminate unused code)
- TypeScript/JSX transpilation
- CSS processing, asset hashing
- Source maps

**What "lazy build" could mean**:

```
Option A: Build on first request (like Next.js on-demand compilation)
  → Request comes in → compile just the requested route on-demand
  → Cache compiled output → serve subsequent requests from cache
  → Pro: zero build time on startup
  → Con: first-request latency, cold start on deploy

Option B: No-build dev server extended to production
  → Serve source files via Vite's transform pipeline at runtime
  → Pro: truly zero build step
  → Con: significantly slower response times, no tree shaking

Option C: Pre-compiled modules (import maps + native ESM)
  → Serve ES modules directly, browser resolves imports
  → Pro: no bundler needed
  → Con: waterfall requests on initial load, large apps get slow
```

Vite's plugin system with `@vitejs/plugin-legacy` does something in this space. Astro's Content Collections do on-demand compilation. The "dev experience in production" dream requires solving the waterfall problem.

The most promising direction I've seen: build-on-deploy with cached incremental compilation (only rebuild what changed), making the build step fast enough that "no build" is less valuable. Turbopack/Rspack are going this direction.

What specific pain is the lazy build solving — CI time, iteration speed, or something else?
